receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,512 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snmpreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/snmpreceiver"
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/gosnmp/gosnmp"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+)
+
+type oidDataType byte
+
+// Trimmed down the larger list of gosnmp data types to make retrieving data a
+// little more simple when using this client
+const (
+	notSupportedVal oidDataType = 0x00
+	integerVal      oidDataType = 0x01 // value will be int64
+	floatVal        oidDataType = 0x02 // value will be float64
+	stringVal       oidDataType = 0x03 // value will be string
+)
+
+// snmpData used for processFunc and is a simpler version of gosnmp.SnmpPDU
+type snmpData struct {
+	parentOID string // optional
+	oid       string
+	value     interface{}
+	valueType oidDataType
+}
+
+// processFunc our own function type to better control what data can be passed in
+type processFunc func(data snmpData) error
+
+// client is used for retrieving data from a SNMP environment
+type client interface {
+	// GetScalarData retrieves SNMP scalar data from a list of passed in OIDS,
+	// then the passed in function is performed on each piece of data
+	GetScalarData(oids []string, processFn processFunc, scraperErrors *scrapererror.ScrapeErrors)
+	// GetIndexedData retrieves SNMP indexed data from a list of passed in OIDS,
+	// then the passed in function is performed on each piece of data
+	GetIndexedData(oids []string, processFn processFunc, scraperErrors *scrapererror.ScrapeErrors)
+	// Connect makes a connection to the SNMP host
+	Connect() error
+	// Close closes a connection to the SNMP host
+	Close() error
+}
+
+// snmpClient implements the client interface and retrieves data through SNMP
+type snmpClient struct {
+	client goSNMPWrapper
+	logger *zap.Logger
+}
+
+// Verify snmpClient implements client interface
+var _ client = (*snmpClient)(nil)
+
+// newClient creates an initialized client
+// Relies on config being validated thoroughly
+func newClient(cfg *Config, _ component.Host, _ component.TelemetrySettings, logger *zap.Logger) (client, error) {
```

Comment:
Why pass in Host and TelemetrySettings if they are not used?

Reply:
@djaglowski good callout. oversight and will fix